### PR TITLE
Seebs/ergodoxled

### PR DESCRIPTION
**Pull request by seebs**
Sadly, I can't make these faster. But I can make them COOLER. Since we have to write both sets of LEDs separately anyway, we might as well make them more awesome, and allow driving them separately. So in ERGODOX_LED_30 mode, you can have a single rainbow covering both halves, the "knight" mode bounces from one side to the other, and so on.

### CODE REVIEW

File: keyboards/ergodox_ez/keymaps/default/keymap.c [hunk 1]
```diff
@@ -194,6 +194,8 @@ bool process_record_user(uint16_t keycode, keyrecord_t *record) {
 
 // Runs just one time when the keyboard initializes.
 void matrix_init_user(void) {
+  rgblight_init();
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/default/keymap.c]: Not needed.  This happens regardless, as long as the RGBLights are enabled.

**Comment by seebs** [comment]
Hmm. I am actually unclear on when it's needed. There's a convention in a couple of other configs of checking for RGBLIGHT_COLOR_LAYER_0, and if it's not set, calling rgblight_init(). I suppose that's to reset to a default mode which might have been set? Hmm.

**Comment by drashna** [comment]
_init is called here:
https://github.com/qmk/qmk_firmware/blob/3e2fd642798cde75acbecd99750e8b70eab52537/tmk_core/common/keyboard.c#L151

But the _enable should only be needed to ensure that the layer indication is activated.  And technically, it's only needed once, ever.   Otherwise, the layer indication may not work until the lights are enabled.  (but that may be desirable behavior)

**Comment by Taudris** [comment]
#2350 moves the I2C LED code into a separate driver from the WS2812 code, which allows sending different LED data to the two halves. If it gets merged, it would allow you to do this in `rgblight_set()`, which would be much cleaner and would enable you to make the defines not specific to ErgoDox.

**Comment by drashna** [comment]
Actually, there have been some changes to the split keyboard code that may help with this. Such as "rgbled_split"

File: drivers/avr/ws2812.c [hunk 2]
```diff
@@ -171,12 +171,24 @@ void inline ws2812_setleds_rgbw(LED_TYPE *ledarray, uint16_t leds)
...
     I2C_Start();
     I2C_Write(0x84);
-    uint16_t datlen = leds<<2;
-    uint8_t curbyte;
-    uint8_t * data = (uint8_t*)ledarray;
-    while (datlen--) {
-      curbyte=*data++;
-      I2C_Write(curbyte);
+    int i = 0;
+    #if defined(ERGODOX_LED_30)
```
Comment (yanfali) [inline drivers/avr/ws2812.c]: is there a different way to do this? what if you made this function weak and just overrode it in the ergodox keyboard? Feels wrong to have keyboard specific code in a driver.

Comment (seebs) [reply drivers/avr/ws2812.c]: I agree that this is awful, but the actual driver itself seems to also be innately keyboard-specific, so far as I know -- it's specific to the ergodox EZ's weird i2c -> GPIO expander situation.

File: keyboards/ergodox_ez/config.h [hunk 3]
```diff
@@ -78,14 +94,20 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+
 #define RGBW 1
 
 /* Set 0 if debouncing isn't needed */
 #define DEBOUNCE    15
 
 #define PREVENT_STUCK_MODIFIERS
 
-#define USB_MAX_POWER_CONSUMPTION 500
+#define USB_MAX_POWER_CONSUMPTION 400
```
Comment (skullydazed) [inline keyboards/ergodox_ez/config.h]: Does this change have something to do with the rest of the PR?

Comment (seebs) [reply keyboards/ergodox_ez/config.h]: i don't think it did, and i honestly have no idea how that happened.

File: keyboards/ergodox_ez/keymaps/default/keymap.c (hunk 1 above)
Comment (drashna) [reply keyboards/ergodox_ez/keymaps/default/keymap.c]: Also, now we have `keyboard_post_init_user` that happens after rgblight initialization occurs

**Comment by drashna** [comment]
It looks like there is a merge conflict here.

If you could fix this?  Otherwise, is it okay if we just go ahead and close this?

**Comment by seebs** [comment]
Rebased. The merge conflicts were all spacing cleanup in the code. Like, `leds<<2` vs. `leds << 2`.

File: keyboards/ergodox_ez/config.h (hunk 3 above)
Comment (drashna) [reply keyboards/ergodox_ez/config.h]: Could this be reverted then? 
```suggestion
#define USB_MAX_POWER_CONSUMPTION 500
```

**Comment by seebs** [comment]
wtf. I looked for that change in my tree, and didn't see it... but now of course I do.

cleaned it up and force-pushed in the hopes that it'll go away.

**Comment by drashna** [comment]
Oh may, yeah, git can be ... funny, some times. 

Let me test how this works, first. If it is more that a couple of days, ping me, plz.

**Comment by drashna** [comment]
Testing this out .... it does not work right....  Basically, it HAS to be set at the keyboard level, and it never gets picked up on the user level.  

Eg, in my layout, if I add the defines, it gets picked up by the driver, but not the keyboard.  So it uses 30 code, but for 15 LEDs. 

Some of this would need to be changed/fixed.  But I'm not exactly sure how.

**Comment by seebs** [comment]
Huh. I just checked, and on my keyboard, it looks like I'm getting the mirrored LED behavior, which may or may not be what I thought I had it configured for.

I'll see if I can free up some time to look at this again and figure out what I thought it was doing.

Thinking about it, though: I don't think I ever tried to make this work inside a keymap, I think it had to be in the keyboard's config.h, which is why there were any patches at all to config.h.

**Comment by drashna** [comment]
post_config.h should work for this, actually.

**Comment by drashna** [comment]
Do you mind if I take this over, and add some additional config stuff, based on the newer code (such as above)

**Comment by seebs** [comment]
not at all, i'm clearly not keeping up with things.

**Comment by drashna** [comment]
Awesome!  And I'd rather ask permission. 

And sorry that this never got merged, but I'll make sure that you have credit for the new code (since it's mostly based on this)

**Pull request closed without merging**